sdl-deltalake/src/test/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObjectTest.scala

```diff
@@ -320,4 +323,144 @@ class DeltaLakeTableDataObjectTest extends FunSuite with BeforeAndAfter {
+    val targetTable = Table(db = Some("default"), name = "test_inc", primaryKey = Some(Seq("id")))
+    val targetTablePath = tempPath+s"/${targetTable.fullName}"
+    val targetDO = DeltaLakeTableDataObject("deltaDO1", table = targetTable, path=Some(targetTablePath), saveMode = SDLSaveMode.Append)
+    targetDO.dropTable
+    targetDO.setState(None) // initialize incremental output with empty state
+
+    // write test data 1
+    val df1 = Seq((1, "A", 1), (2, "A", 2), (3, "B", 3), (4, "B", 4)).toDF("id", "p", "value")
+    targetDO.prepare
+    targetDO.initSparkDataFrame(df1, Seq())
+    targetDO.writeSparkDataFrame(df1)
+    val newState1 = targetDO.getState
+
+    // test 1
+    targetDO.setState(newState1)
+    targetDO.getSparkDataFrame()(contextExec).count() shouldEqual 4
+
+
+    // append test data 2
+    val df2 = Seq((5, "B", 5)).toDF("id", "p", "value")
+    targetDO.writeSparkDataFrame(df2)
+    val newState2 = targetDO.getState
+
+    // test 2
+    targetDO.setState(newState2)
+    targetDO.getSparkDataFrame()(contextExec).count() shouldEqual 1
+
+    // append test data 3
+    val df3 = Seq((6, "T", 5), (7, "R", 7), (8, "T", 2)).toDF("id", "p", "value")
+    targetDO.writeSparkDataFrame(df3)
+    val newState3 = targetDO.getState
+
+    // test 3
+    targetDO.setState(newState3)
+    targetDO.getSparkDataFrame()(contextExec).count() shouldEqual 3
+
+    assert(newState1.get < newState2.get)
+    assert(newState2.get < newState3.get)
+
+    targetDO.setState(None) // to get the full dataframe
+    targetDO.getSparkDataFrame()(contextInit).count() shouldEqual 8
+  }
+
+  test("incremental output mode without primary keys") {
+
+    // create data object
+    val targetTable = Table(db = Some("default"), name = "test_inc")
+    val targetTablePath = tempPath + s"/${targetTable.fullName}"
+    val targetDO = DeltaLakeTableDataObject("deltaDO1", table = targetTable, path = Some(targetTablePath), saveMode = SDLSaveMode.Append)
+    targetDO.dropTable
+    targetDO.setState(None) // initialize incremental output with empty state
+
+    // write test data
+    val df1 = Seq((1, "A", 1), (2, "A", 2), (3, "B", 3), (4, "B", 4)).toDF("id", "p", "value")
+    targetDO.prepare
+    targetDO.initSparkDataFrame(df1, Seq())
+    targetDO.writeSparkDataFrame(df1)
+    val newState1 = targetDO.getState
+    targetDO.setState(newState1)
+    targetDO.getSparkDataFrame()(contextExec).count() shouldEqual 4
+
+    val df2 = Seq((5, "B", 5)).toDF("id", "p", "value")
+    targetDO.writeSparkDataFrame(df2)
+    val newState2 = targetDO.getState
+
+    // test
+    val thrown = intercept[IllegalArgumentException] {
+      targetDO.setState(newState2)
+      targetDO.getSparkDataFrame()(contextExec).count()
+    }
+
+    // check
+    assert(thrown.isInstanceOf[IllegalArgumentException])
+    assert(thrown.getMessage == s"requirement failed: PrimaryKey for table [${targetDO.table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
```

Comment:
I would simplify the check (i just added $(id) to the message in a previous comment...)
```suggestion
    assert(thrown.getMessage.contains("PrimaryKey for table"))
```